gossip/hub/handlers.go

```diff
@@ -0,0 +1,664 @@
+	return &root, nil
+}
+
+func getSTH(ctx context.Context, c *hubInfo, w http.ResponseWriter, r *http.Request) (int, error) {
+	root, err := GetLogRoot(ctx, c.rpcClient, c.logID, c.hubPrefix)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	hth := api.HubTreeHead{
+		TreeSize:  root.TreeSize,
+		Timestamp: root.TimestampNanos, // in nanoseconds since UNIX epoch
+		RootHash:  root.RootHash,
+	}
+	hthData, err := tls.Marshal(hth)
+	if err != nil {
+		return http.StatusInternalServerError, fmt.Errorf("failed to marshal tree head: %v", err)
+	}
+	signature, err := c.signer.Sign(hthData)
+	if err != nil {
+		return http.StatusInternalServerError, fmt.Errorf("failed to sign STH data: %v", err)
+	}
+
+	// Now build the final result object and marshal to JSON
+	jsonRsp := api.GetSTHResponse{TreeHeadData: hthData, HubSignature: signature}
+	jsonData, err := json.Marshal(&jsonRsp)
+	if err != nil {
+		return http.StatusInternalServerError, fmt.Errorf("failed to marshal response: %v %v", jsonRsp, err)
+	}
+
+	w.Header().Set(contentTypeHeader, contentTypeJSON)
+	if _, err := w.Write(jsonData); err != nil {
+		return http.StatusInternalServerError, fmt.Errorf("failed to write response data: %v", err)
+	}
+
+	return http.StatusOK, nil
+}
+
+func getSTHConsistency(ctx context.Context, c *hubInfo, w http.ResponseWriter, r *http.Request) (int, error) {
+	firstVal := r.FormValue(api.GetSTHConsistencyFirst)
+	if firstVal == "" {
+		return http.StatusBadRequest, errors.New("parameter 'first' is required")
+	}
+	secondVal := r.FormValue(api.GetSTHConsistencySecond)
+	if secondVal == "" {
+		return http.StatusBadRequest, errors.New("parameter 'second' is required")
+	}
+
+	first, err := strconv.ParseInt(firstVal, 10, 64)
+	if err != nil {
+		return http.StatusBadRequest, errors.New("parameter 'first' is malformed")
+	}
+	second, err := strconv.ParseInt(secondVal, 10, 64)
+	if err != nil {
+		return http.StatusBadRequest, errors.New("parameter 'second' is malformed")
+	}
+
+	if first < 0 || second < 0 {
+		return http.StatusBadRequest, fmt.Errorf("first and second params cannot be <0: %d %d", first, second)
+	}
+	if second < first {
+		return http.StatusBadRequest, fmt.Errorf("invalid first, second params: %d %d", first, second)
+	}
+
+	var jsonRsp api.GetSTHConsistencyResponse
+	if first != 0 {
+		req := trillian.GetConsistencyProofRequest{LogId: c.logID, FirstTreeSize: first, SecondTreeSize: second}
+
+		glog.V(2).Infof("%s: GetSTHConsistency(%d, %d) => grpc.GetConsistencyProof %+v", c.hubPrefix, first, second, req)
+		rsp, err := c.rpcClient.GetConsistencyProof(ctx, &req)
+		glog.V(2).Infof("%s: GetSTHConsistency <= grpc.GetConsistencyProof err=%v", c.hubPrefix, err)
+		if err != nil {
+			return c.toHTTPStatus(err), fmt.Errorf("backend GetConsistencyProof request failed: %v", err)
+		}
+
+		// TODO(drysdale): configure the hub with the Trillian public key for this tree, and check the Trillian signature here.
```

Comment:
TODO: replace with `client.UpdateRoot` ?

Reply:
No, I don't think so -- as above, `UpdateRoot` is doing more/different work than is needed.  (The `trillian/client` code is doing a bunch of checks and caching, but the Hub is a passthrough and so doesn't need to -- if it did, it would just be duplicating check that the Hub client/monitor does.)